Tests/src/InternalTests/CommonTypeTests/Optional.cpp

```diff
@@ -0,0 +1,290 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#if !defined(SKIP_INTERNAL_TESTS) || defined(RUN_COMMONTYPE_TESTS) || defined(RUN_COMMONTYPE_OPTIONAL_TESTS)
+
+	#include "CSP/Common/Optional.h"
+
+	#include "CSP/Memory/DllAllocator.h"
+	#include "TestHelpers.h"
+
+	#include <gtest/gtest.h>
+
+
+using namespace csp::common;
+
+class OptionalExtraTestClass;
+
+class OptionalTestClass
+{
+public:
+	OptionalTestClass(const int SomeFieldValue) : SomeField(SomeFieldValue) {};
+
+	int SomeField;
+
+	int CopyCount = 0;
+	int MoveCount = 0;
+
+	bool Moved = false;
+
+	OptionalTestClass(const OptionalTestClass& Other)
+	{
+		SomeField = Other.SomeField;
+		CopyCount = Other.CopyCount + 1;
+		MoveCount = Other.MoveCount;
+	}
+
+	OptionalTestClass(OptionalTestClass&& Other)
+	{
+		SomeField	= Other.SomeField;
+		CopyCount	= Other.CopyCount;
+		MoveCount	= Other.MoveCount + 1;
+		Other.Moved = true;
+	}
+
+	OptionalTestClass(OptionalTestClass* Other)
+	{
+		SomeField = Other->SomeField;
+	}
+};
+
+class OptionalExtraTestClass : public OptionalTestClass
+{
+public:
+	OptionalExtraTestClass(const int SomeFieldValue, const int ExtraFieldValue) : OptionalTestClass(SomeFieldValue), ExtraField(ExtraFieldValue) {};
+
+	int ExtraField;
+};
+
+CSP_INTERNAL_TEST(CSPEngine, CommonOptionalTests, OptionalDefaultInitialisationTest)
+{
+	try
+	{
+		Optional<OptionalTestClass> OptionalInstance;
+
+		EXPECT_FALSE(OptionalInstance.HasValue());
+		EXPECT_EQ(&(*OptionalInstance), nullptr);
+	}
+	catch (...)
+	{
+		FAIL();
+	}
+}
+
+CSP_INTERNAL_TEST(CSPEngine, CommonOptionalTests, OptionalNullptrInitialisationTest)
+{
+	try
+	{
+		Optional<OptionalTestClass> OptionalInstance(nullptr);
+
+		EXPECT_FALSE(OptionalInstance.HasValue());
+		EXPECT_EQ(&(*OptionalInstance), nullptr);
+	}
+	catch (...)
+	{
+		FAIL();
+	}
+}
+
+CSP_INTERNAL_TEST(CSPEngine, CommonOptionalTests, OptionalWithPointerInitialisationTest)
+{
+
+	try
+	{
+		auto* TestClassInstancePtr = (OptionalTestClass*) csp::memory::DllAlloc(sizeof(OptionalTestClass));
+		new (TestClassInstancePtr) OptionalTestClass(2);
+
+		Optional<OptionalTestClass> OptionalInstance(TestClassInstancePtr);
+
+		EXPECT_TRUE(OptionalInstance.HasValue());
+		EXPECT_EQ(&(*OptionalInstance), TestClassInstancePtr);
+		EXPECT_EQ((*OptionalInstance).SomeField, TestClassInstancePtr->SomeField);
+		EXPECT_EQ(OptionalInstance->SomeField, TestClassInstancePtr->SomeField);
+	}
+	catch (...)
+	{
+		FAIL();
+	}
+}
+
+CSP_INTERNAL_TEST(CSPEngine, CommonOptionalTests, OptionalWithPointerAndDestructorInitialisationTest)
+{
+	try
+	{
+		bool DestructorRan = false;
+
+		std::function<void(OptionalTestClass*)> CustomDestructor = [&](OptionalTestClass* Pointer)
+		{
+			Pointer->~OptionalTestClass();
+			csp::memory::DllFree(Pointer);
+			DestructorRan = true;
+		};
+
+		{
+			auto* TestClassInstancePtr = (OptionalTestClass*) csp::memory::DllAlloc(sizeof(OptionalTestClass));
```

Comment:
I didn't use `CSP_NEW` as it is not available since it is defined in `Memory.h` which is not within our include folder. 

Could switch to using `new`/`delete`.

Reply:
`<Memory.h>` is accessible in the tests project as we directly include the CSP source code.

Reply:
Hmm, dunno what I did before when I tried it but it's working for now alright.